### DISCUSSION TITLE
How to get the field name from a UniqueConstraintViolationException

### DISCUSSION THREAD

**Discussion by squivix** [mikro-orm/mikro-orm#3689] [Q&A] [answered]
I'm trying to extract the field name from a UniqueConstraintViolationException, and although serializing the error seems to yield plenty of information that includes the field name:

```json
{
    "length": 211,
    "name": "UniqueConstraintViolationException",
    "severity": "ERROR",
    "code": "23505",
    "detail": "Key (username)=(johndoe) already exists.",
    "schema": "public",
    "table": "user",
    "constraint": "user_username_unique",
    "file": "nbtinsert.c",
    "line": "663",
    "routine": "_bt_check_unique"
}
```

The typescript type `UniqueConstraintViolationException` does not include anything particularly helpful, so I can't get it at compile time without casting to any.

Is there any other type with more specific info? Possibly one that is driver specific, or do I have to cast to any?

**Comment by B4nan** [comment, marked as answer]
This is an error thrown by the database, it's not coming from the ORM (neither from any other JS libraries, you can see the file that throws is in fact in C) - the ORM only rethrows it with a specific error type for easier detection - that is the `UniqueConstraintViolationException` class. 

> The typescript type UniqueConstraintViolationException does not include anything particularly helpful, so I can't get it at compile time without casting to any.

Not sure what you mean by that.

**Reply by squivix** [reply]
> Not sure what you mean by that.

I mean that a lot of the information available at runtime about the error that is not defined in the type `UniqueConstraintViolationException`.

For example, if I wanted to get the `detail` property from the error, which I know exists at runtime:
![image](https://user-images.githubusercontent.com/55770995/199975245-e1d47e2e-e241-45e3-b0a3-482fe58831bb.png)

I can't, because it is not defined in `UniqueConstraintViolationException` 

![image](https://user-images.githubusercontent.com/55770995/199975632-94a6b4f0-2fda-4acd-8159-1866d3b9e99d.png)

I would have to cast to any.

**Reply by B4nan** [reply]
> I mean that a lot of the information available at runtime about the error that is not defined in the type UniqueConstraintViolationException.

Not really, that information you are up to is only at the database level - once again, the error is thrown by your database, not via any JS library on top of it. The SQL needs to be interpreted to know it (maybe even executed, not just parsed). The ORM is not doing that anyhow - it is a non-goal.

> For example, if I wanted to get the detail property from the error, which I know exists at runtime:

The type is shared for all drivers, this `detail` property is driver specific. So yes, you will need to cast it.
